### PR TITLE
Added support for the "crop" method.

### PR DESCRIPTION
**Pull request by clickworkorange**
I had a need for this functionality, and while I know I could use the `vips!` method in my uploader it felt like an omission - so I put together a PR with a (passing) spec. The test is a little basic, but I wasn't sure if it would be ok to add new image fixtures, which I would need to do in order to test that the cropping produces the correct output (my test only checks the resulting dimensions). I'd be happy to look at implementing further vips methods if desired!

### CODE REVIEW

**Comment by clickworkorange** [comment]
Two tests are failing, but I'm pretty sure neither has anything to do with me! 

**1. Test / RSpec and Cucumber (ruby-head, gemfiles/rails-7-1.gemfile, false) (pull_request)** 
````
LoadError:
       cannot load such file -- RMagick (You may need to install the rmagick gem)
     # ./lib/carrierwave/processing/rmagick.rb:66:in `rescue in block in <module:RMagick>'
     # ./lib/carrierwave/processing/rmagick.rb:62:in `block in <module:RMagick>'
     # ./vendor/bundle/ruby/3.4.0+0/gems/activesupport-7.1.3/lib/active_support/concern.rb:138:in `class_eval'
     # ./vendor/bundle/ruby/3.4.0+0/gems/activesupport-7.1.3/lib/active_support/concern.rb:138:in `append_features'
     # ./spec/orm/activerecord_spec.rb:182:in `include'
     # ./spec/orm/activerecord_spec.rb:182:in `block (5 levels) in <top (required)>'
     # ./vendor/bundle/ruby/3.4.0+0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
     # ./vendor/bundle/ruby/3.4.0+0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
     # ./vendor/bundle/ruby/3.4.0+0/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # ./vendor/bundle/ruby/3.4.0+0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
     # ./vendor/bundle/ruby/3.4.0+0/gems/webmock-3.19.1/lib/webmock/rspec.rb:39:in `block (2 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # LoadError:
     #   cannot load such file -- observer
     #   ./vendor/bundle/ruby/3.4.0+0/gems/rmagick-5.3.0/lib/rmagick_internal.rb:24:in `<top (required)>'
````
**2. Test / RSpec and Cucumber (jruby-head, gemfiles/rails-7-0.gemfile, true) (pull_request)**
````
CarrierWave::ActiveRecord#mount_uploader #image should return valid XML when to_xml is called when image is nil
     Failure/Error: hash = Hash.from_xml(@event.to_xml)["event"]

     REXML::ParseException:
       #<RuntimeError: Illegal character "\u0000" in raw string "true\"/>\u0000\u0000\u0000">
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/text.rb:140:in `block in check'
       org/jruby/RubyArray.java:1989:in `each'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/text.rb:136:in `check'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/attribute.rb:175:in `element='
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/element.rb:2384:in `[]='
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/parsers/treeparser.rb:36:in `block in parse'
       org/jruby/RubyHash.java:1610:in `each'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/parsers/treeparser.rb:35:in `parse'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/document.rb:448:in `build'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rexml-3.2.6/lib/rexml/document.rb:101:in `initialize'
       org/jruby/RubyClass.java:921:in `new'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/activesupport-7.0.8/lib/active_support/xml_mini/rexml.rb:29:in `parse'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/activesupport-7.0.8/lib/active_support/xml_mini.rb:95:in `parse'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/activesupport-7.0.8/lib/active_support/core_ext/hash/conversions.rb:152:in `initialize'
       org/jruby/RubyClass.java:934:in `new'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/activesupport-7.0.8/lib/active_support/core_ext/hash/conversions.rb:129:in `from_xml'
       /home/runner/work/carrierwave/carrierwave/spec/orm/activerecord_spec.rb:94:in `block in <main>'
       org/jruby/RubyBasicObject.java:2620:in `instance_exec'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:263:in `block in run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:511:in `block in with_around_and_singleton_context_hooks'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:468:in `block in with_around_example_hooks'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:486:in `block in run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:352:in `call'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
       org/jruby/RubyKernel.java:1707:in `loop'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block in setup'
       org/jruby/RubyBasicObject.java:2620:in `instance_exec'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:457:in `instance_exec'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:390:in `execute_with'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:628:in `block in run_around_example_hooks_for'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:352:in `call'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/webmock-3.19.1/lib/webmock/rspec.rb:39:in `block in <main>'
       org/jruby/RubyBasicObject.java:2620:in `instance_exec'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:457:in `instance_exec'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:390:in `execute_with'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:628:in `block in run_around_example_hooks_for'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:352:in `call'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:629:in `run_around_example_hooks_for'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/hooks.rb:486:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:468:in `with_around_example_hooks'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:511:in `with_around_and_singleton_context_hooks'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example.rb:259:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example_group.rb:646:in `block in run_examples'
       org/jruby/RubyArray.java:2811:in `map'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example_group.rb:651:in `run_examples'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example_group.rb:607:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example_group.rb:608:in `block in run'
       org/jruby/RubyArray.java:2811:in `map'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example_group.rb:608:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example_group.rb:608:in `block in run'
       org/jruby/RubyArray.java:2811:in `map'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/example_group.rb:608:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/runner.rb:121:in `block in run_specs'
       org/jruby/RubyArray.java:2811:in `map'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/runner.rb:121:in `block in run_specs'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/configuration.rb:2070:in `with_suite_hooks'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/runner.rb:116:in `block in run_specs'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/reporter.rb:74:in `report'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/runner.rb:115:in `run_specs'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/runner.rb:89:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/runner.rb:71:in `run'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/lib/rspec/core/runner.rb:45:in `invoke'
       /home/runner/work/carrierwave/carrierwave/vendor/bundle/jruby/3.1.0/gems/rspec-core-3.12.2/exe/rspec:4:in `<main>'
       ...
       Illegal character "\u0000" in raw string "true\"/>\u0000\u0000\u0000"
       Line: 3
       Position: 80
````
Trust Java to produce an error that's over a mile long because of an illegal character... :laughing:

**Comment by clickworkorange** [comment]
Only one test failing now, on JRuby-Head. Been looking at it for a while and it seems to have to do with the serialisation of ActiveRecord object; it's failing on `@event.to_xml` every time, with `@event` being an ActiveRecord object: 

```` ruby
# activerecord_spec.rb

describe CarrierWave::ActiveRecord do
  def create_table(name)
    ActiveRecord::Base.connection.create_table(name, force: true) do |t|
      t.column :image, :string
      t.column :images, :json
      t.column :textfile, :string
      t.column :textfiles, :json
      t.column :foo, :string
    end
  end
 
 ...

  def reset_class(class_name)
    Object.send(:remove_const, class_name) rescue nil
    Object.const_set(class_name, Class.new(ActiveRecord::Base))
  end

  before(:all) { create_table("events") }
  after(:all) { drop_table("events") }

  before do
    @uploader = Class.new(CarrierWave::Uploader::Base)
    reset_class("Event")
    @event = Event.new
  end

  ...

      it "should return valid XML when to_xml is called when image is nil" do
        #### THIS TEST FAILS ON @event.to_xml ####
        hash = Hash.from_xml(@event.to_xml)["event"]

        expect(@event[:image]).to be_nil
        expect(hash.keys).to include("image")
        expect(hash["image"].keys).to include("url")
        expect(hash["image"]["url"]).to be_nil
      end

      it "should return valid XML when to_xml is called when image is present" do
        #### THIS TEST ALSO FAILS ON @event.to_xml ####
        @event[:image] = 'test.jpeg'
        @event.save!
        @event.reload

        expect(Hash.from_xml(@event.to_xml)["event"]["image"]).to eq({"url" => "/uploads/test.jpeg"})
      end
      
      ...
````
It might be that a change in `rexml` and/or `activemodel-serializers-xml` which pulls it in is where the "illegal" null characters are coming from (\u0000) - but whatever that change might be it *only* appears to affect JRuby.

**Comment by clickworkorange** [comment]
Some things which might be relevant: 

- `activemodel-serializers-xml` hasn't seen any changes for the last four years, so that's not it. 
- `Rexml` had a new release on July 23 - but since it's `rexml` complaining about those null characters my feeling is that the problem lies elsewhere, namely: 
- `JRuby` had [a major release in November](https://github.com/jruby/jruby/releases/tag/9.4.5.0) last year, which added Ruby 3.1 compatibility. I'm really not a Java guy (*>shudder<*), but this looks like it *may* be relevant: 

https://github.com/jruby/jruby/commit/830c0dc0f0c916bbbcad0ad001d0a288f93350d0

**Comment by clickworkorange** [comment]
I've added a question to the JRuby commit mentioned in my previous comment: 

https://github.com/jruby/jruby/commit/830c0dc0f0c916bbbcad0ad001d0a288f93350d0

**Comment by mshibuya** [comment]
Thank you so much for taking a deep look on the JRuby failure!
It's not a blocker for this PR anyway, as it's unrelated.

But is it possible to add `#crop` to `CarrierWave::MiniMagick` and `CarrierWave::RMagick` as well? They should be considered as a sort of adapter, so they'd better to have the same interface unless doing so is infeasible.

**Comment by clickworkorange** [comment]
I'll take a look!

**Comment by clickworkorange** [comment]
I have added support for image cropping to the RMagic and MiniMagic processors as well. In doing so I found a difference in behaviour between Vips and the other two: while both RMagic and MiniMagic silently accept cropping beyond the image boundaries (retaining the original image bottom/right edge), Vips would throw an `Vips::Error: extract_area: bad extract area` exception when asked to do so. I thought it made most sense to make Vips match the behaviour of the other two processors, and did so by recalculating the cropping area if either edge of it falls outside the image: 

```` ruby
def crop(left, top, width, height, combine_options: {})
  width, height = resolve_dimensions(width, height)
  width = vips_image.width - left if width + left > vips_image.width
  height = vips_image.height - top if height + top > vips_image.height

  vips! do |builder|
    builder.crop(left, top, width, height)
      .apply(combine_options)
  end
end
````
I added specs to all three processors to test for this case.

**Comment by clickworkorange** [comment]
I'm working on updating the README to include information about Vips support and the `crop` method - would you prefer that as a separate PR, or should I include those changes here?

**Comment by mshibuya** [comment]
You can include it into this PR 👍

**Comment by clickworkorange** [comment]
Ok, here's what I had in mind: 

----

## Manipulating images
[no changes]

### Using MiniMagick
[no changes]

### Using RMagick
[no changes]

### Using Vips

CarrierWave version 2.2.0 added support for the `libvips` image processing library, through [ImageProcessing::Vips](https://github.com/janko/image_processing/blob/master/doc/vips.md). Its functionality matches that of the RMagic and MiniMagic processors, but it uses less memory and offers [faster processing](https://github.com/libvips/libvips/wiki/Speed-and-memory-use). To use the Vips processing module you must first install `libvips`, for example: 

````bash
$ sudo apt install libvips
````

You also need to tell your uploader to use Vips:

````ruby
class ImageFileUploader < CarrierWave::Uploader::Base
  include CarrierWave::Vips
  
  ...

end
````

### List of available processing methods:

> [!NOTE]
> While the intetion is to provide uniform interfaces to al three processing libraries the availability and implementation of processing methods can <a href="supported-processing-methods">vary slightly between them</a>.

- `convert` - Changes the image encoding format to the given format (eg. jpg). This operation is treated specially to trigger the change of the file extension, so it matches with the format of the resulting file.
- `resize_to_limit` - Resize the image to fit within the specified dimensions while retaining the original aspect ratio. Will only resize the image if it is larger than the specified dimensions. The resulting image may be shorter or narrower than specified in the smaller dimension but will not be larger than the specified values.
- `resize_to_fit` - Resize the image to fit within the specified dimensions while retaining the original aspect ratio. The image may be shorter or narrower than specified in the smaller dimension but will not be larger than the specified values.
- `resize_to_fill` - Resize the image to fit within the specified dimensions while retaining the aspect ratio of the original image. If necessary, crop the image in the larger dimension. Optionally, a "gravity" may be specified, for example "Center", or "NorthEast".
- `resize_and_pad` - Resize the image to fit within the specified dimensions while retaining the original aspect ratio. If necessary, will pad the remaining area with the given color, which defaults to transparent (for gif and png, white for jpeg). Optionally, a "gravity" may be specified, as above.
- `crop` - Crop the image to the contents of a box with the specified height and width, positioned a given number of pixels from the top and left. The original image edge will be retained should the bottom and/or right edge of the box fall outside the image bounds. 

#### Supported processing methods

The following table shows which processing methods are supported by each processing library, and which parameters they accept: 

Method|RMagick|MiniMagick|Vips
------|-----------------|-----------------|-----------------|
`convert`|`format`|`format`, `page`<sup>1</sup>|`format`, `page`<sup>1</sup>
`resize_to_limit`|`width`, `height`|`width`, `height`|`width`, `height`
`resize_to_fit`|`width`, `height`|`width`, `height`|`width`, `height`
`resize_to_fill`|`width`, `height`, `gravity`<sup>2</sup>|`width`, `height`, `gravity`<sup>2</sup>|`width`, `height`
`resize_and_pad`|`width`, `height`, `background`, `gravity`<sup>2</sup>|`width`, `height`, `background`, `gravity`<sup>2</sup>|`width`, `height`, `background`, `gravity`<sup>2</sup>
`resize_to_geometry_string`|`geometry_string`<sup>3</sup>|not implemented|not implemented
`crop`|`left`, `top`, `width`, `height`|`left`, `top`, `width`, `height`|`left`, `top`, `width`, `height`

<sup>1</sup>`page` refers to the page number when converting from PDF, frame number when converting from GIF, and layer number when converting from PSD.

<sup>2</sup>`gravity` refers to an image position given as one of `Center`, `North`, `NorthWest`, `West`, `SouthWest`, `South`, `SouthEast`, `East`, or `NorthEast`.

<sup>3</sup>`geometry_string` is an [ImageMagick geometry string](https://rmagick.github.io/imusage.html#geometry).

File: README.md [hunk 1]
```diff
@@ -1018,6 +1010,54 @@ end
 Check out the manipulate! method, which makes it easy for you to write your own
 manipulation methods.
 
+### Using Vips
+
+CarrierWave version 2.2.0 added support for the `libvips` image processing library, through [ImageProcessing::Vips](https://github.com/janko/image_processing/blob/master/doc/vips.md). Its functionality matches that of the RMagic and MiniMagic processors, but it uses less memory and offers [faster processing](https://github.com/libvips/libvips/wiki/Speed-and-memory-use). To use the Vips processing module you must first install `libvips`, for example:
```
Comment (mshibuya) [inline README.md:1015]: Found a little typo:

```suggestion
CarrierWave version 2.2.0 added support for the `libvips` image processing library, through [ImageProcessing::Vips](https://github.com/janko/image_processing/blob/master/doc/vips.md). Its functionality matches that of the RMagick and MiniMagick processors, but it uses less memory and offers [faster processing](https://github.com/libvips/libvips/wiki/Speed-and-memory-use). To use the Vips processing module you must first install `libvips`, for example: 
```

File: README.md [hunk 2]
```diff
@@ -1018,6 +1010,54 @@ end
...
+`resize_to_fill`|`width`, `height`, `gravity`<sup>2</sup>|`width`, `height`, `gravity`<sup>2</sup>|`width`, `height`
+`resize_and_pad`|`width`, `height`, `background`, `gravity`<sup>2</sup>|`width`, `height`, `background`, `gravity`<sup>2</sup>|`width`, `height`, `background`, `gravity`<sup>2</sup>
+`resize_to_geometry_string`|`geometry_string`<sup>3</sup>|*not implemented*|*not implemented*
+`crop`|`left`, `top`, `width`, `height`|`left`, `top`, `width`, `height`|`left`, `top`, `width`, `height`
+
+<sup>1</sup>`page` refers to the page number when converting from PDF, frame number when converting from GIF, and layer number when converting from PSD.
+
+<sup>2</sup>`gravity` refers to an image position given as one of `Center`, `North`, `NorthWest`, `West`, `SouthWest`, `South`, `SouthEast`, `East`, or `NorthEast`.
+
+<sup>3</sup>`geometry_string` is an [ImageMagick geometry string](https://rmagick.github.io/imusage.html#geometry).
```
Comment (mshibuya) [inline README.md:1059]: Nice comparison table 👍

**Comment by mshibuya** [comment]
Great work, thank you so much!

**Comment by clickworkorange** [comment]
My pleasure!

**Pull request merged**